### PR TITLE
feat: Add non-entity retrieval support for ClickHouse offline store

### LINKED ISSUE #5835: ClickHouse - historical retrieval without entity dataframe
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by YassinNouh21**
# What this PR does / why we need it:

Adds support for non-entity historical retrieval (`entity_df=None`) in the ClickHouse offline store, bringing it to parity with the PostgreSQL offline store.

**Changes:**
- Updated `ClickhouseOfflineStore.get_historical_features()` to accept `entity_df=None` with optional `start_date`/`end_date` kwargs
- When `entity_df` is `None`, a synthetic single-row DataFrame is created using the provided date range (or sensible defaults: `end_date=now`, `start_date` derived from max TTL or 30 days)
- Added 3 unit tests covering: both dates provided, end_date only (start from TTL), and no dates (defaults to now)

**Usage:**
```python
fs.get_historical_features(
    features=["driver_stats:conv_rate"],
    entity_df=None,
    start_date=datetime(2023, 1, 1),
    end_date=datetime(2023, 6, 1),
)
```

# Which issue(s) this PR fixes:

Fixes #5835

# Test plan

- [x] Unit tests pass (`pytest tests/unit/infra/offline_stores/test_clickhouse.py`)
- [x] Ruff lint and format checks pass
- [x] E2E verification of full FeatureStore -> Provider -> ClickhouseOfflineStore chain
- [x] No regression on existing entity_df mode
- [x] DCO sign-off included

---

<a href="https://app.devin.ai/review/feast-dev/feast/pull/6066" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

File: sdk/python/feast/infra/offline_stores/contrib/clickhouse_offline_store/clickhouse.py [hunk 1]
```diff
@@ -43,15 +44,42 @@ def get_historical_features(
...
+                if max_ttl_seconds > 0:
+                    start_date = end_date - timedelta(seconds=max_ttl_seconds)
+                else:
+                    start_date = end_date - timedelta(days=30)
+            else:
+                start_date = make_tzaware(start_date)
+
+            entity_df = pd.DataFrame({"event_timestamp": [end_date]})
+
         entity_schema = _get_entity_schema(entity_df, config)
```
[the remark this answers is not in the document]
Comment (YassinNouh21) [reply sdk/python/feast/infra/offline_stores/contrib/clickhouse_offline_store/clickhouse.py:67]: this is not correct

[the remark this answers is not in the document]
Comment (ntkathole) [reply sdk/python/feast/infra/offline_stores/contrib/clickhouse_offline_store/clickhouse.py:67]: @YassinNouh21 if start_date is not needed, either we can remove the start_date computation entirely since it's dead code ? or can use start_date in the entity_df for consistency with other stores?

File: sdk/python/tests/unit/infra/offline_stores/test_clickhouse.py [hunk 2]
```diff
@@ -133,3 +134,211 @@ def test_clickhouse_config_handles_none_additional_client_args():
...
+            timestamp_field="event_timestamp",
+        ),
+        schema=[
+            Field(name="feature1", dtype=Float32),
+            Field(name="feature2", dtype=Float32),
+        ],
+    )
+
+
+class TestNonEntityRetrieval:
```
Comment (ntkathole) [inline sdk/python/tests/unit/infra/offline_stores/test_clickhouse.py:139]: Tests seems over-mocked

File: sdk/python/feast/infra/offline_stores/contrib/clickhouse_offline_store/clickhouse.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (YassinNouh21) [reply sdk/python/feast/infra/offline_stores/contrib/clickhouse_offline_store/clickhouse.py:67]: Removed dead start_date computation and simplified tests (shared helper, 3 → 2 tests). Fixed in latest commit.

**Comment by franciscojavierarceo** [comment]
Why not have an integration test?

File: sdk/python/tests/unit/infra/offline_stores/test_clickhouse.py (hunk 2 above)
Comment (YassinNouh21) [reply sdk/python/tests/unit/infra/offline_stores/test_clickhouse.py:139]: You're right, removed the heavy mocking. For proper coverage, an integration test against a real ClickHouse instance would be more valuable than over-mocked unit tests — the unit tests here may not be necessary at all.

**Comment by YassinNouh21** [comment]
@franciscojavierarceo @ntkathole take another look !

File: sdk/python/feast/infra/offline_stores/contrib/clickhouse_offline_store/clickhouse.py [hunk 3]
```diff
@@ -43,15 +44,26 @@ def get_historical_features(
...
 
+        # Handle non-entity retrieval mode
+        if entity_df is None:
+            end_date = kwargs.get("end_date", None)
+            if end_date is None:
+                end_date = _utc_now()
+            else:
+                end_date = make_tzaware(end_date)
+
+            entity_df = pd.DataFrame({"event_timestamp": [end_date]})
```
[the remark this answers is not in the document]
Comment (ntkathole) [reply sdk/python/feast/infra/offline_stores/contrib/clickhouse_offline_store/clickhouse.py:65]: @YassinNouh21 This seems critical issue.

[the remark this answers is not in the document]
Comment (YassinNouh21) [reply sdk/python/feast/infra/offline_stores/contrib/clickhouse_offline_store/clickhouse.py:65]: Not a bug — this is intentional. The PIT join uses MAX(entity_timestamp) as the upper bound, so the timestamp in the synthetic entity_df IS the query upper bound. Using [end_date] gives the window [end_date - TTL, end_date], which is correct. The Postgres implementation using pd.date_range(start=start_date, ...)[:1] actually has the bug — it takes start_date as the sole timestamp, making end_date unreachable. Our implementation matches Dask and is the correct behavior.
@ntkathole

[the remark this answers is not in the document]
Comment (franciscojavierarceo) [reply sdk/python/feast/infra/offline_stores/contrib/clickhouse_offline_store/clickhouse.py:65]: @ntkathole sounds like there's a bug in postgres???

[the remark this answers is not in the document]
Comment (YassinNouh21) [reply sdk/python/feast/infra/offline_stores/contrib/clickhouse_offline_store/clickhouse.py:65]: Yes I noticed that

[the remark this answers is not in the document]
Comment (ntkathole) [reply sdk/python/feast/infra/offline_stores/contrib/clickhouse_offline_store/clickhouse.py:65]: It will get resolved via https://github.com/feast-dev/feast/pull/6057 cc @Vperiodt

Comment (YassinNouh21) [inline sdk/python/feast/infra/offline_stores/contrib/clickhouse_offline_store/clickhouse.py:65]: Good call using `[end_date]` here. The Postgres implementation at `postgres.py:162-168` has a subtle bug where it uses:
```python
pd.date_range(start=start_date, end=end_date, freq="1s", tz=timezone.utc)[:1]
```
`[:1]` takes the **first** element which is `start_date`, so the PIT join upper bound (`MAX(entity_timestamp)`) becomes `start_date` instead of `end_date` — making data after `start_date` unreachable.

This implementation correctly uses `end_date` as the entity timestamp, matching Dask's behavior and ensuring the PIT window is `[end_date - TTL, end_date]`.

File: sdk/python/feast/infra/offline_stores/contrib/clickhouse_offline_store/clickhouse.py [hunk 4]
```diff
@@ -43,15 +44,26 @@ def get_historical_features(
...
         full_feature_names: bool = False,
+        **kwargs,
     ) -> RetrievalJob:
         assert isinstance(config.offline_store, ClickhouseOfflineStoreConfig)
         for fv in feature_views:
             assert isinstance(fv.batch_source, ClickhouseSource)
 
+        # Handle non-entity retrieval mode
+        if entity_df is None:
+            end_date = kwargs.get("end_date", None)
```
Comment (YassinNouh21) [inline sdk/python/feast/infra/offline_stores/contrib/clickhouse_offline_store/clickhouse.py:59]: Only `end_date` is extracted from kwargs, but `start_date` is not. The Postgres store supports both `start_date` and `end_date` and has a dedicated `{% if start_date and end_date %}` branch in its query template for range-based retrieval.

Without `start_date` support, users can't restrict how far back the PIT join looks (beyond TTL). This means this isn't full parity with Postgres — it's single-point retrieval at `end_date`, not range-based retrieval over `[start_date, end_date]`.

Consider extracting and forwarding `start_date` to the query builder in a follow-up, and adding a non-entity branch to the ClickHouse query template.

File: sdk/python/feast/infra/offline_stores/contrib/clickhouse_offline_store/clickhouse.py [hunk 5]
```diff
@@ -43,15 +44,26 @@ def get_historical_features(
         config: RepoConfig,
         feature_views: List[FeatureView],
         feature_refs: List[str],
-        entity_df: Union[pd.DataFrame, str],
+        entity_df: Optional[Union[pd.DataFrame, str]],
         registry: BaseRegistry,
         project: str,
         full_feature_names: bool = False,
+        **kwargs,
```
Comment (YassinNouh21) [inline sdk/python/feast/infra/offline_stores/contrib/clickhouse_offline_store/clickhouse.py:51]: nit: Since only `end_date` is used from `**kwargs`, consider accepting `end_date` as an explicit keyword argument instead of relying on `**kwargs`. This makes the API self-documenting and easier to type-check:
```python
def get_historical_features(
    ...,
    full_feature_names: bool = False,
    end_date: Optional[datetime] = None,
) -> RetrievalJob:
```
Though if `start_date` support is planned, keeping `**kwargs` for now is fine.

File: sdk/python/tests/unit/infra/offline_stores/test_clickhouse.py [hunk 6]
```diff
@@ -133,3 +134,109 @@ def test_clickhouse_config_handles_none_additional_client_args():
...
+        )
+        assert isinstance(job, RetrievalJob)
+
+    def test_non_entity_mode_defaults_end_date(self):
+        """entity_df=None without end_date defaults to now."""
+        from feast.infra.offline_stores.offline_store import RetrievalJob
+
+        fv = self._make_feature_view("test_fv")
+        job = self._call_get_historical_features([fv])
+        assert isinstance(job, RetrievalJob)
```
Comment (YassinNouh21) [inline sdk/python/tests/unit/infra/offline_stores/test_clickhouse.py:242]: The unit tests verify that a `RetrievalJob` is returned, but don't assert anything about the generated SQL or the synthetic entity_df contents. Consider adding an assertion that the entity_df passed to `_upload_entity_df` contains the expected `end_date` timestamp — this would catch regressions like the Postgres `start_date` bug.

Also, a test for `entity_df=None` **without** any `end_date` kwarg would verify the `_utc_now()` default path.
